{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8026763-c935-475a-b71b-a50f363b3c91",
   "metadata": {},
   "source": [
    "# scRNA-seq and scATAC-seq integration analysis\n",
    "2026.02.20 @carushi "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ae2efd",
   "metadata": {
    "vscode": {
     "languageId": "bat"
    }
   },
   "source": [
    "## Preprocess\n",
    "* Download scRNA-seq and scATAC-seq metadata from GEO\n",
    "    * https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE273197\n",
    "* scRNA-seq data\n",
    "    * GSE273197_scrna_X.mtx.gz (normalized)\n",
    "    * GSE273197_scrna_obs.csv.gz\n",
    "    * GSE273197_scrna_var.csv.gz\n",
    "    * GSE273197_scrna_X_raw.mtx.gz (raw count)\n",
    "    * GSE273197_scrna_obs_raw.csv.gz\n",
    "    * GSE273197_scrna_var_raw.csv.gz\n",
    "* scATAC-seq data\n",
    "    * GSE273197_scatac_snap_obj_clust.rds.gz\n",
    "* Integrated data\n",
    "    * GSE273197_scrna_atac_integrated.h5Seurat\n",
    "* Place the files under data/geo_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cec585a9",
   "metadata": {
    "vscode": {
     "languageId": "bat"
    }
   },
   "source": [
    "## Preprocess 2\n",
    "* Get whitelist data provided by 10x Genomics\n",
    "    * scRNA-seq: 3M-february-2018.txt.gz\n",
    "    * scATAC-seq: 737K-cratac-v1.txt\n",
    "    * See https://kb.10xgenomics.com/s/article/115004506263-What-is-a-barcode-inclusion-list-formerly-barcode-whitelist\n",
    "* Download CellMarker 2.0 annotation file\n",
    "    * Open Cell_marker_All.xlsx using Excel\n",
    "    * Save as txt file\n",
    "    * nkf -Lu all_cell_marker.txt > all_cell_marker_lu.txt (for UNIX)\n",
    "    * mv all_cell_marker_lu.txt ../data/gene/\n",
    "    * run parse_cell_marker.R\n",
    "* Download Mat (gene set) files\n",
    "    * Download from xxxxxx and plate them under /data/gene/ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7552cd93-125e-4d9f-9328-024d7d638654",
   "metadata": {},
   "outputs": [],
   "source": [
    "grep \"\\\"CD[0-9]\" 16-90.merged.gtf | grep \"\\tgene\\t\" | cut -d\"\\t\" -f 9- | cut -d\" \" -f6 | sed 's/\\\"//g' | sed 's/;//g' > cd_gene.txt\n",
    "grep chrM 16-90.merged.gtf | grep \"\\ttranscript\" | cut -d\"\\t\" -f9 | cut -d\" \" -f2 | sed s/\\\"//g | sed s/\\;//g > chrM_gene.txt\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "785dd525",
   "metadata": {},
   "source": [
    "## scRNA-seq analysis (original)\n",
    "* Run STARsolo for alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b716395",
   "metadata": {
    "vscode": {
     "languageId": "bat"
    }
   },
   "outputs": [],
   "source": [
    "# bash ../scripts/alignment/mapping/alignment_solo.sh\n",
    "# bash ../scripts/alignment/mapping/alignment_solo_refseq.sh # Refseq genes\n",
    "# bash ../scripts/alignment/mapping/alignment_solo_quad.sh   # use 16-90 personal genome annotation \n",
    "bash ../scripts/alignment/mapping/alignment_solo_cd4.sh    # change CD4 annotation from Ensembl to Refseq -> Used for downstream analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ffb6dd7-3b2f-4938-b431-94290cb88a84",
   "metadata": {
    "tags": []
   },
   "source": [
    "## scRNA-seq analysis by Scanpy\n",
    "* Start from here\n",
    "* gunzip ../data/geo_metadata/ files\n",
    "* Run scRNA_scanpy.ipynb (please follow Data setup 2)\n",
    "* scarmadillo_filt_scobj.pyn and scarmadillo_filt_modf_scobj.pyn will be created"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3abcfc6f",
   "metadata": {},
   "source": [
    "## Data conversion\n",
    "* Convert scRNA-seq data into csv and create seurat_scrna_integrated.rds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "575ac34f-d25c-47fc-8121-4842304f52b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "python ../scripts/alignment/mapping/write_scanpy_obj.py scarmadillo_filtered_modf_scobj.pyn scarmadillo_filtered\n",
    "Rscript ../scripts/alignment/mapping/read_scRNAseq.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a7eb0b4-bdd8-467a-ac4d-4af45051d82e",
   "metadata": {},
   "source": [
    "## scATAC-seq analysis (original)\n",
    "### Run STARsolo alignment and SnapTools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4d4b25a-fc21-4f4c-9a53-6edbeb3f3a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "bash ../scripts/alignment/mapping/alignment_scatac_by_star.sh\n",
    "# bash ../scripts/alignment/mapping/alignment_scatac_by_star_relaxed.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4af81f8f-1545-4897-b631-cfdc16022e2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "bash ../scripts/alignment/snaptools_preprocess.sh\n",
    "# bash ../scripts/alignment/snaptools_preprocess_pub.sh\n",
    "# bash ../scripts/alignment/snaptools_preprocess_relaxed.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cef1414a-02bc-4135-a9cb-af55ecbb47e7",
   "metadata": {},
   "source": [
    "## scATAC-seq analysis by SnapATAC\n",
    "### Set up conda environments\n",
    "* genome - insntall MACS2, snapatac, and HOMER at ~/miniconda/env/genome/bin\n",
    "* You can change this setting defined in multi_sc_base.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cad20eca-1d47-4188-834f-69b9e1500201",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript ../scripts/sc_clustering/snap_basic_stats.r make_list file_list.txt\n",
    "Rscript ../scripts/sc_clustering/snap_basic_stats.r plot file_list.txt 500 2.35"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e9ce507-89e1-4faa-8607-d9b9712132f6",
   "metadata": {},
   "source": [
    "### Relaxed setting\n",
    "abolished"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa1f39f-bdd1-4bbb-ab4f-6159fab67948",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rscript ../scripts/sc_clustering/snap_basic_stats.r make_list file_list.txt\n",
    "# Rscript ../scripts/sc_clustering/snap_basic_stats.r plot file_list.txt 1500 2.35"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bee85e1c-d81d-430e-8721-15ba47704b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter out unreliable cells\n",
    "Rscript ../scripts/sc_clustering/snap_atac_clustering.R snap_obj_filtered.rds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f49c7c6-4671-4ca5-b4da-8fce6296f6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript ../scripts/sc_clustering/snap_obj_analysis.R snap_obj_clust.rds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddfd494a-696e-4c6d-8b25-c53f5bec6980",
   "metadata": {},
   "source": [
    "## Seurat CCA integration and cell type annotation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b14c18db-4cab-4ea0-9ee0-4737f1baa575",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Clustering and annotation\n",
    "Rscript ../scripts/sc_clustering/coembedding_for_cell_typing.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dc57318-e3c0-4ad4-9415-c80c76c85294",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript ../scripts/sc_clustering/seurat_obj_analysis.R final_pseudo_bulk_atac.rds\n",
    "Rscript ../scripts/sc_clustering/seurat_obj_analysis.R final_pseudo_bulk_rna.rds\n",
    "Rscript ../scripts/sc_clustering/seurat_obj_analysis.R final_coembed.rds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de19d137-fe49-4580-ba8e-9515e756f03b",
   "metadata": {},
   "source": [
    "## Extra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eece32a-bfb6-4333-9a6e-be15030faf11",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript extract_barcode.R\n",
    "bash run_cellsnp.sh"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
