{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8026763-c935-475a-b71b-a50f363b3c91",
   "metadata": {},
   "source": [
    "# scRNA-seq and scATAC-seq integration analysis\n",
    "2026.02.20 @carushi "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ae2efd",
   "metadata": {
    "vscode": {
     "languageId": "bat"
    }
   },
   "source": [
    "## Preprocess\n",
    "* Download scRNA-seq and scATAC-seq metadata from GEO\n",
    "    * https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE273197\n",
    "* scRNA-seq data (normalized)\n",
    "    * GSE273197_scrna_X.mtx.gz (normalized)\n",
    "    * GSE273197_scrna_obs.csv.gz\n",
    "    * GSE273197_scrna_var.csv.gz\n",
    "* scRNA-seq data (raw count)\n",
    "    * GSE273197_scrna_X_raw.mtx.gz\n",
    "    * GSE273197_scrna_obs_raw.csv.gz\n",
    "    * GSE273197_scrna_var_raw.csv.gz\n",
    "* scATAC-seq data\n",
    "    * GSE273197_scatac_snap_obj_clust.rds.gz\n",
    "* Integrated data\n",
    "    * GSE273197_scrna_atac_integrated.h5Seurat\n",
    "* Place the files under data/geo_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cec585a9",
   "metadata": {
    "vscode": {
     "languageId": "bat"
    }
   },
   "source": [
    "## Preprocess 2\n",
    "* Get whitelist data provided by 10x Genomics\n",
    "    * scRNA-seq: 3M-february-2018.txt.gz\n",
    "    * scATAC-seq: 737K-cratac-v1.txt\n",
    "    * See https://kb.10xgenomics.com/s/article/115004506263-What-is-a-barcode-inclusion-list-formerly-barcode-whitelist\n",
    "* Download CellMarker 2.0 annotation file\n",
    "    * Open Cell_marker_All.xlsx using Excel\n",
    "    * Save as txt file\n",
    "    * nkf -Lu all_cell_marker.txt > all_cell_marker_lu.txt (for UNIX)\n",
    "    * mv all_cell_marker_lu.txt ../data/gene/\n",
    "    * run parse_cell_marker.R\n",
    "* Download Mat (gene set) files\n",
    "    * Download from xxxxxx and plate them under /data/gene/ "
   ]
  },
  {
   "cell_type": "raw",
   "id": "10606bc2-c161-4c68-b1be-7b4a78bf719e",
   "metadata": {},
   "source": [
    "grep \"\\\"CD[0-9]\" 16-90.merged.gtf | grep \"\\tgene\\t\" | cut -d\"\\t\" -f 9- | cut -d\" \" -f6 | sed 's/\\\"//g' | sed 's/;//g' > cd_gene.txt\n",
    "grep chrM 16-90.merged.gtf | grep \"\\ttranscript\" | cut -d\"\\t\" -f9 | cut -d\" \" -f2 | sed s/\\\"//g | sed s/\\;//g > chrM_gene.txt\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "785dd525",
   "metadata": {},
   "source": [
    "## scRNA-seq analysis (original)\n",
    "* Run STARsolo for alignment"
   ]
  },
  {
   "cell_type": "raw",
   "id": "53ade623-2d6c-4413-9e26-23311dc41a1a",
   "metadata": {
    "vscode": {
     "languageId": "bat"
    }
   },
   "source": [
    "# bash ../scripts/alignment/mapping/alignment_solo.sh\n",
    "# bash ../scripts/alignment/mapping/alignment_solo_refseq.sh # Refseq genes\n",
    "# bash ../scripts/alignment/mapping/alignment_solo_quad.sh   # use 16-90 personal genome annotation \n",
    "bash ../scripts/alignment/mapping/alignment_solo_cd4.sh    # change CD4 annotation from Ensembl to Refseq -> Used for downstream analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ffb6dd7-3b2f-4938-b431-94290cb88a84",
   "metadata": {
    "tags": []
   },
   "source": [
    "## scRNA-seq analysis by Scanpy\n",
    "* Start from here\n",
    "* gunzip ../data/geo_metadata/ files\n",
    "* Run scRNA_scanpy.ipynb (please follow Data setup 2)\n",
    "* scarmadillo_filt_scobj.pyn and scarmadillo_filt_modf_scobj.pyn will be created"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3abcfc6f",
   "metadata": {},
   "source": [
    "## Data conversion\n",
    "* Convert scRNA-seq data into csv and create seurat_scrna_integrated.rds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "320cddd1-333c-404e-ae7f-fde0b515929c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 要求されたパッケージ SeuratObject をロード中です \n",
      " 要求されたパッケージ sp をロード中です \n",
      "\n",
      " 次のパッケージを付け加えます: ‘SeuratObject’ \n",
      "\n",
      " 以下のオブジェクトは ‘package:base’ からマスクされています:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      " 警告メッセージ: \n",
      "1:  パッケージ ‘Seurat’ はバージョン 4.3.3 の R の下で造られました  \n",
      "2:  パッケージ ‘SeuratObject’ はバージョン 4.3.3 の R の下で造られました  \n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h [1]  0  3  6  2  5  8 14  7  4 10  9 11 13 12  1 15\n",
      "\u001b[?25h\u001b[?25h[1] \"../data/bam/bam_p_cd4/305264Solo.out/Gene/filtered\"\n",
      "[1] 33472  4855\n",
      "[1] 33472     2\n",
      "[1] 4855   11\n",
      "[1] 4855    1\n",
      "                V1          barcode batch  cov n_counts_spike n_counts\n",
      "1 AAACCCACAGCACCCA AAACCCACAGCACCCA     0 3151              0     3149\n",
      "2 AAACCCACATGACAAA AAACCCACATGACAAA     0 1999              0     1999\n",
      "3 AAACCCAGTCTGTCAA AAACCCAGTCTGTCAA     0 6498              0     6497\n",
      "4 AAACCCATCCACGTGG AAACCCATCCACGTGG     0 2628              0     2627\n",
      "5 AAACGAAAGACCATAA AAACGAAAGACCATAA     0 3841              0     3840\n",
      "6 AAACGAAAGTTAGAAC AAACGAAAGTTAGAAC     0 7060              0     7060\n",
      "      tsne1       tsne2     umap1     umap2 cluster_louvain cluster_leiden\n",
      "1 -36.37129  -0.8910956 18.850536 11.871915               1              0\n",
      "2 -14.78877 -35.2076340 11.315924 11.853340               3              3\n",
      "3  18.08184 -11.9773840 -1.852669  5.478721               0              6\n",
      "4  34.71863 -13.4296610 -3.080211  7.933577               0              6\n",
      "5  34.37264 -15.9756800 -3.118085  7.707276               0              6\n",
      "6 -31.41229   1.6131216 18.134739 11.923431               1              0\n",
      "           barcode batch  cov n_counts_spike n_counts     tsne1       tsne2\n",
      "1 AAACCCACAGCACCCA     0 3151              0     3149 -36.37129  -0.8910956\n",
      "2 AAACCCACATGACAAA     0 1999              0     1999 -14.78877 -35.2076340\n",
      "3 AAACCCAGTCTGTCAA     0 6498              0     6497  18.08184 -11.9773840\n",
      "4 AAACCCATCCACGTGG     0 2628              0     2627  34.71863 -13.4296610\n",
      "5 AAACGAAAGACCATAA     0 3841              0     3840  34.37264 -15.9756800\n",
      "6 AAACGAAAGTTAGAAC     0 7060              0     7060 -31.41229   1.6131216\n",
      "      umap1     umap2 cluster_louvain cluster_leiden\n",
      "1 18.850536 11.871915               1              0\n",
      "2 11.315924 11.853340               3              3\n",
      "3 -1.852669  5.478721               0              6\n",
      "4 -3.080211  7.933577               0              6\n",
      "5 -3.118085  7.707276               0              6\n",
      "6 18.134739 11.923431               1              0\n",
      "             gene_id gene_name\n",
      "1 ENSDNOG00000042816      MTDH\n",
      "2 ENSDNOG00000043683   RF00026\n",
      "3 ENSDNOG00000018035    TSPYL5\n",
      "4 ENSDNOG00000036192       CPQ\n",
      "5 ENSDNOG00000027861   RF00003\n",
      "6 ENSDNOG00000038783      SDC2\n",
      " 警告:  Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
      " 警告:  Data is of class dgTMatrix. Coercing to dgCMatrix.\n",
      "[1] \"../data/bam/bam_p_cd4/305265Solo.out/Gene/filtered\"\n",
      "[1] 33472  3110\n",
      "[1] 33472     2\n",
      "[1] 3110   11\n",
      "[1] 3110    1\n",
      "                   V1          barcode batch  cov n_counts_spike n_counts\n",
      "4856 AAACCCAAGAGATTCA AAACCCAAGAGATTCA     1 3862              0     3861\n",
      "4857 AAACCCAAGCGCAATG AAACCCAAGCGCAATG     1 4419              0     4419\n",
      "4858 AAACCCATCGAGAAGC AAACCCATCGAGAAGC     1 5087              0     5086\n",
      "4859 AAACGAAAGACAAGCC AAACGAAAGACAAGCC     1 4832              0     4831\n",
      "4860 AAACGAAAGGGTTGCA AAACGAAAGGGTTGCA     1 1148              0     1148\n",
      "4861 AAACGCTTCGATACGT AAACGCTTCGATACGT     1 3681              0     3681\n",
      "         tsne1      tsne2     umap1     umap2 cluster_louvain cluster_leiden\n",
      "4856 -34.36132  10.631392 18.057991 11.124228               1              0\n",
      "4857  13.55922 -38.373726 10.359496 13.955519               2              2\n",
      "4858  29.92796   9.229588 -0.729463  5.253018               0              1\n",
      "4859 -28.18400  17.461636 16.665690  9.324507               8              8\n",
      "4860 -36.81123  10.711075 18.297705 10.727304               1              0\n",
      "4861 -24.29363  -3.999080 16.691795 11.638087               4              4\n",
      "           barcode batch  cov n_counts_spike n_counts     tsne1       tsne2\n",
      "1 AAACCCACAGCACCCA     0 3151              0     3149 -36.37129  -0.8910956\n",
      "2 AAACCCACATGACAAA     0 1999              0     1999 -14.78877 -35.2076340\n",
      "3 AAACCCAGTCTGTCAA     0 6498              0     6497  18.08184 -11.9773840\n",
      "4 AAACCCATCCACGTGG     0 2628              0     2627  34.71863 -13.4296610\n",
      "5 AAACGAAAGACCATAA     0 3841              0     3840  34.37264 -15.9756800\n",
      "6 AAACGAAAGTTAGAAC     0 7060              0     7060 -31.41229   1.6131216\n",
      "      umap1     umap2 cluster_louvain cluster_leiden\n",
      "1 18.850536 11.871915               1              0\n",
      "2 11.315924 11.853340               3              3\n",
      "3 -1.852669  5.478721               0              6\n",
      "4 -3.080211  7.933577               0              6\n",
      "5 -3.118085  7.707276               0              6\n",
      "6 18.134739 11.923431               1              0\n",
      "             gene_id gene_name\n",
      "1 ENSDNOG00000042816      MTDH\n",
      "2 ENSDNOG00000043683   RF00026\n",
      "3 ENSDNOG00000018035    TSPYL5\n",
      "4 ENSDNOG00000036192       CPQ\n",
      "5 ENSDNOG00000027861   RF00003\n",
      "6 ENSDNOG00000038783      SDC2\n",
      " 警告:  Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
      " 警告:  Data is of class dgTMatrix. Coercing to dgCMatrix.\n",
      "[1] \"../data/bam/bam_p_cd4/305266Solo.out/Gene/filtered\"\n",
      "[1] 33472  3488\n",
      "[1] 33472     2\n",
      "[1] 3488   11\n",
      "[1] 3488    1\n",
      "                   V1          barcode batch  cov n_counts_spike n_counts\n",
      "7966 AAACCCAAGAGTCAAT AAACCCAAGAGTCAAT     2 4671              0     4671\n",
      "7967 AAACCCAAGTAATACG AAACCCAAGTAATACG     2 3313              0     3312\n",
      "7968 AAACCCACACACAGCC AAACCCACACACAGCC     2 2537              0     2536\n",
      "7969 AAACCCACATCATCTT AAACCCACATCATCTT     2 3187              0     3184\n",
      "7970 AAACGAATCGTCAACA AAACGAATCGTCAACA     2 3762              0     3757\n",
      "7971 AAACGCTCACCGCTAG AAACGCTCACCGCTAG     2 5834              0     5834\n",
      "          tsne1      tsne2     umap1      umap2 cluster_louvain cluster_leiden\n",
      "7966  -7.327014 -20.849386 13.483824 13.0524460               3              3\n",
      "7967  -1.343831  21.765810  3.548619 18.4255030               5              5\n",
      "7968 -11.462139  -7.188200 14.696607 11.7107020               4              4\n",
      "7969   2.715722  49.265570  7.382692 -0.7972289               6              7\n",
      "7970 -18.466501 -11.799191 14.700054 12.4950410               4              4\n",
      "7971  18.570564  -1.526404 -0.106813  7.4644300               0              1\n",
      "           barcode batch  cov n_counts_spike n_counts     tsne1       tsne2\n",
      "1 AAACCCACAGCACCCA     0 3151              0     3149 -36.37129  -0.8910956\n",
      "2 AAACCCACATGACAAA     0 1999              0     1999 -14.78877 -35.2076340\n",
      "3 AAACCCAGTCTGTCAA     0 6498              0     6497  18.08184 -11.9773840\n",
      "4 AAACCCATCCACGTGG     0 2628              0     2627  34.71863 -13.4296610\n",
      "5 AAACGAAAGACCATAA     0 3841              0     3840  34.37264 -15.9756800\n",
      "6 AAACGAAAGTTAGAAC     0 7060              0     7060 -31.41229   1.6131216\n",
      "      umap1     umap2 cluster_louvain cluster_leiden\n",
      "1 18.850536 11.871915               1              0\n",
      "2 11.315924 11.853340               3              3\n",
      "3 -1.852669  5.478721               0              6\n",
      "4 -3.080211  7.933577               0              6\n",
      "5 -3.118085  7.707276               0              6\n",
      "6 18.134739 11.923431               1              0\n",
      "             gene_id gene_name\n",
      "1 ENSDNOG00000042816      MTDH\n",
      "2 ENSDNOG00000043683   RF00026\n",
      "3 ENSDNOG00000018035    TSPYL5\n",
      "4 ENSDNOG00000036192       CPQ\n",
      "5 ENSDNOG00000027861   RF00003\n",
      "6 ENSDNOG00000038783      SDC2\n",
      " 警告:  Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
      " 警告:  Data is of class dgTMatrix. Coercing to dgCMatrix.\n",
      "[1] \"../data/bam/bam_p_cd4/305267Solo.out/Gene/filtered\"\n",
      "[1] 33472  4640\n",
      "[1] 33472     2\n",
      "[1] 4640   11\n",
      "[1] 4640    1\n",
      "                    V1          barcode batch  cov n_counts_spike n_counts\n",
      "11454 AAACCCACATCAACCA AAACCCACATCAACCA     3 4909              0     4900\n",
      "11455 AAACCCACATGACGGA AAACCCACATGACGGA     3 1921              0     1920\n",
      "11456 AAACCCACATGGGATG AAACCCACATGGGATG     3 6084              0     6080\n",
      "11457 AAACCCATCCCACAGG AAACCCATCCCACAGG     3 4793              0     4793\n",
      "11458 AAACGAAAGCGAGTCA AAACGAAAGCGAGTCA     3 1659              0     1658\n",
      "11459 AAACGAAAGCGTTACT AAACGAAAGCGTTACT     3 1743              0     1742\n",
      "          tsne1      tsne2       umap1     umap2 cluster_louvain cluster_leiden\n",
      "11454 30.209518  41.771870  3.58633070  1.370088              12             15\n",
      "11455 -1.992005 -26.922752 12.34200900 12.535757               3              3\n",
      "11456 15.994193  12.387959  0.95419750  5.749461               0              1\n",
      "11457 22.299410   8.738604  0.08453788  6.729867               0              1\n",
      "11458 38.082462 -35.574303  6.15217350 13.363680               9             11\n",
      "11459 34.891865 -35.147316  6.51374440 13.553823               9             11\n",
      "           barcode batch  cov n_counts_spike n_counts     tsne1       tsne2\n",
      "1 AAACCCACAGCACCCA     0 3151              0     3149 -36.37129  -0.8910956\n",
      "2 AAACCCACATGACAAA     0 1999              0     1999 -14.78877 -35.2076340\n",
      "3 AAACCCAGTCTGTCAA     0 6498              0     6497  18.08184 -11.9773840\n",
      "4 AAACCCATCCACGTGG     0 2628              0     2627  34.71863 -13.4296610\n",
      "5 AAACGAAAGACCATAA     0 3841              0     3840  34.37264 -15.9756800\n",
      "6 AAACGAAAGTTAGAAC     0 7060              0     7060 -31.41229   1.6131216\n",
      "      umap1     umap2 cluster_louvain cluster_leiden\n",
      "1 18.850536 11.871915               1              0\n",
      "2 11.315924 11.853340               3              3\n",
      "3 -1.852669  5.478721               0              6\n",
      "4 -3.080211  7.933577               0              6\n",
      "5 -3.118085  7.707276               0              6\n",
      "6 18.134739 11.923431               1              0\n",
      "             gene_id gene_name\n",
      "1 ENSDNOG00000042816      MTDH\n",
      "2 ENSDNOG00000043683   RF00026\n",
      "3 ENSDNOG00000018035    TSPYL5\n",
      "4 ENSDNOG00000036192       CPQ\n",
      "5 ENSDNOG00000027861   RF00003\n",
      "6 ENSDNOG00000038783      SDC2\n",
      " 警告:  Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
      " 警告:  Data is of class dgTMatrix. Coercing to dgCMatrix.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "!python ../scripts/alignment/miscs/write_scanpy_obj.py scarmadillo_filtered_modf_scobj.pyn scarmadillo_filtered\n",
    "!Rscript ../scripts/alignment/miscs/read_scRNAseq.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a7eb0b4-bdd8-467a-ac4d-4af45051d82e",
   "metadata": {},
   "source": [
    "## scATAC-seq analysis (original)\n",
    "### Run STARsolo alignment and SnapTools\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b4a1a35a-487b-49b9-830c-3f6bcc0fa31b",
   "metadata": {},
   "source": [
    "bash ../scripts/alignment/mapping/alignment_scatac_by_star.sh\n",
    "# bash ../scripts/alignment/mapping/alignment_scatac_by_star_relaxed.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4af81f8f-1545-4897-b631-cfdc16022e2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls: ../data/bam_pub_genome//*.out.bam: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!bash ../scripts/alignment/snaptools_preprocess.sh\n",
    "# bash ../scripts/alignment/snaptools_preprocess_pub.sh\n",
    "# bash ../scripts/alignment/snaptools_preprocess_relaxed.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cef1414a-02bc-4135-a9cb-af55ecbb47e7",
   "metadata": {},
   "source": [
    "## scATAC-seq analysis by SnapATAC\n",
    "### Set up conda environments\n",
    "* genome - insntall MACS2, snapatac, and HOMER at ~/miniconda/env/genome/bin\n",
    "* You can change this setting defined in multi_sc_base.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cad20eca-1d47-4188-834f-69b9e1500201",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript ../scripts/sc_clustering/snap_basic_stats.r make_list file_list.txt\n",
    "Rscript ../scripts/sc_clustering/snap_basic_stats.r plot file_list.txt 500 2.35"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e9ce507-89e1-4faa-8607-d9b9712132f6",
   "metadata": {},
   "source": [
    "### Relaxed setting\n",
    "abolished"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa1f39f-bdd1-4bbb-ab4f-6159fab67948",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rscript ../scripts/sc_clustering/snap_basic_stats.r make_list file_list.txt\n",
    "# Rscript ../scripts/sc_clustering/snap_basic_stats.r plot file_list.txt 1500 2.35"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bee85e1c-d81d-430e-8721-15ba47704b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter out unreliable cells\n",
    "Rscript ../scripts/sc_clustering/snap_atac_clustering.R snap_obj_filtered.rds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f49c7c6-4671-4ca5-b4da-8fce6296f6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript ../scripts/sc_clustering/snap_obj_analysis.R snap_obj_clust.rds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddfd494a-696e-4c6d-8b25-c53f5bec6980",
   "metadata": {},
   "source": [
    "## Seurat CCA integration and cell type annotation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b14c18db-4cab-4ea0-9ee0-4737f1baa575",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Clustering and annotation\n",
    "Rscript ../scripts/sc_clustering/coembedding_for_cell_typing.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dc57318-e3c0-4ad4-9415-c80c76c85294",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript ../scripts/sc_clustering/seurat_obj_analysis.R final_pseudo_bulk_atac.rds\n",
    "Rscript ../scripts/sc_clustering/seurat_obj_analysis.R final_pseudo_bulk_rna.rds\n",
    "Rscript ../scripts/sc_clustering/seurat_obj_analysis.R final_coembed.rds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de19d137-fe49-4580-ba8e-9515e756f03b",
   "metadata": {},
   "source": [
    "## Extra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eece32a-bfb6-4333-9a6e-be15030faf11",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript extract_barcode.R\n",
    "bash run_cellsnp.sh"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
